tools/setup_helpers/ninja_builder.py

```diff
@@ -0,0 +1,81 @@
+import os
+import setuptools
+import distutils
+from contextlib import contextmanager
+
+
+# on the fly create a ninja file in build/ and then
+# run it when run() is called.
+class NinjaBuilder(object):
+    def __init__(self, name):
+        import ninja
+        build_dir = 'build'
+        if not os.path.exists(build_dir):
+            os.mkdir(build_dir)
+        self.name = name
+        self.filename = os.path.join(build_dir, 'build.{}.ninja'.format(name))
+        print('Ninja build file at {}'.format(self.filename))
+        self.writer = ninja.Writer(open(self.filename, 'w'))
+        self.writer.rule('do_cmd', '$cmd')
+
+    def run(self):
+        import ninja
+        self.writer.close()
+        assert(0 == ninja._program('ninja', ['-f', self.filename]))
+        # weird build logic in build develop causes some things to be run
+        # twice so make sure even after we run the command we still
+        # reset this to a valid state
+        # don't use the same name or you can't inspect the real ninja files
+        self.__init__(self.name + "_")
+
+
+class ninja_build_ext(setuptools.command.build_ext.build_ext):
+    def _build_default(self, ext):
+        return setuptools.command.build_ext.build_ext.build_extension(self, ext)
+
+    def build_extension(self, ext):
+        builder = NinjaBuilder(ext.name)
+
+        @contextmanager
+        def patch(obj, attr_name, val):
+            orig_val = getattr(obj, attr_name)
+            setattr(obj, attr_name, val)
+            yield
+            setattr(obj, attr_name, orig_val)
```

Comment:
nit: not terribly important, but this won't run if an exception is thrown in the inner scope. It should be wrapped in a `try: ... finally: ...` block. Just to keep in mind for the future